### PR TITLE
Field content value alerts

### PR DESCRIPTION
**Pull request by lennartkoopmann**
This PR introduces field string value alerts to get an alert whenever at least one message with a field matching a given string has been received since the last alert check run.

Issues:

 * I'm not really happy with the naming of the alert type yet. Suggestions?
 * There are no tests yet. Waiting for #1159 which I cannot fix myself as my DI and MongoDB mocking experience in our codebase is basically non-existent. 

Documentation: Graylog2/documentation@5f61dd22cef89b2496c6611c8fe84f5cdb8d7868

Must be merged together with Graylog2/graylog2-web-interface#1305.

### CODE REVIEW

**Comment by lennartkoopmann** [comment]
Name suggestion: String occurence condition. Maybe @zexxon has some ideas, too?

**Comment by lennartkoopmann** [comment]
Yes, we really need to change the name because it does not only work with strings but also with other data types. :trollface:

File: graylog2-server/src/main/java/org/graylog2/alerts/types/FieldStringValueAlertCondition.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    protected CheckResult runCheck() {
+        String filter = "streams:" + stream.getId();
+        String query = new StringBuilder()
+                .append(field).append(":\"").append(value).append("\"")
+                .toString();
+        Integer backlogSize = getBacklog();
+
+        if(backlogSize == null || backlogSize == 0) {
+            LOG.debug("No backlog size configured. Setting to default 100.");
+            backlogSize = 100;
```
Comment (dennisoelkers) [inline graylog2-server/src/main/java/org/graylog2/alerts/types/FieldStringValueAlertCondition.java]: Why are you defaulting to 100 if the backlog size is 0? a) We are not doing this anywhere else, why is this case different? b) This way it is not possible to define an alarm callback that is triggering an alert without a backlog. c) 100 seems to be a very arbitrary size. Why exactly this number?

File: graylog2-server/src/main/java/org/graylog2/alerts/types/FieldStringValueAlertCondition.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+import org.graylog2.plugin.Message;
+import org.graylog2.plugin.MessageSummary;
+import org.graylog2.plugin.Tools;
+import org.graylog2.plugin.streams.Stream;
+import org.joda.time.DateTime;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.util.Collections;
```
Comment (dennisoelkers) [inline graylog2-server/src/main/java/org/graylog2/alerts/types/FieldStringValueAlertCondition.java]: This one is unused and can be removed.

Comment (lennartkoopmann) [inline graylog2-server/src/main/java/org/graylog2/alerts/types/FieldStringValueAlertCondition.java]: fixed in 90d7154

File: graylog2-server/src/main/java/org/graylog2/alerts/types/FieldStringValueAlertCondition.java (hunk 1 above)
Comment (lennartkoopmann) [inline graylog2-server/src/main/java/org/graylog2/alerts/types/FieldStringValueAlertCondition.java]: Indeed! This came from using the backlog for the search query that needs a limit parameter. Fixed/improved in 8bb355c

**Comment by kroepke** [comment]
I'd rename it to 'Field Content condition'

**Comment by dennisoelkers** [comment]
I also had the idea of naming it "Field Content" condition because it makes more sense. At the same time I think that all three names (Message Count, Field Value, Field Content) are not self-explanatory and slightly confusing. Internally this naming is okay, in the user interface we should use a naming that reflects *what they do* and not *what they are*. For example: "More/less than X messages", "Field has lower/higher value", "Field is exactly".

TLDR; Rename classes and labels to FieldContentAlertCondition/... and improve the naming in the web interface in a subsequent commit/pull request.

**Comment by lennartkoopmann** [comment]
I like Dennis' idea a lot!

**Comment by henrikjohansen** [comment]
My humble suggestion would be "Field metric condition" and "Field string condition" ....

**Comment by lennartkoopmann** [comment]
Renamed it to "Field Content condition" and opened a follow up ticket to rethink the alerts UI: Graylog2/graylog2-web-interface#1330

**Comment by lennartkoopmann** [comment]
The last open thing before this can be merged is the missing tests. @joschi did you find something out about #1159? (See description of this PR)

Thanks so far everybody!

**Comment by lennartkoopmann** [comment]
I added a first simple test for the constructor but failed to write tests for the actual logic in time.

Problems:

* There seems to be no easy way to mock `SearchResult` objects yet.
* Tests are failing because of MongoDB NPEs

Can somebody look at this with me tomorrow after standup? I leave the merging to your own decision and how much sense it makes giving the release time schedule for tomorrow.

**Comment by kroepke** [comment]
Unfortunately this PR is made against master, which does not have the test fixes yet.
I need to pull them in, but really I would appreciate another clean PR after I have ported them.

**Comment by joschi** [comment]
Superseded by #1186.
